README.md

[![Arduino CI](https://github.com/RobTillaart/GAMMA/workflows/Arduino%20CI/badge.svg)](https://github.com/marketplace/actions/arduino_ci)
[![Arduino-lint](https://github.com/RobTillaart/GAMMA/actions/workflows/arduino-lint.yml/badge.svg)](https://github.com/RobTillaart/GAMMA/actions/workflows/arduino-lint.yml)
[![JSON check](https://github.com/RobTillaart/GAMMA/actions/workflows/jsoncheck.yml/badge.svg)](https://github.com/RobTillaart/GAMMA/actions/workflows/jsoncheck.yml)
[![License: MIT](https://img.shields.io/badge/license-MIT-green.svg)](https://github.com/RobTillaart/GAMMA/blob/master/LICENSE)
[![GitHub release](https://img.shields.io/github/release/RobTillaart/GAMMA.svg?maxAge=3600)](https://github.com/RobTillaart/GAMMA/releases)


# GAMMA

Arduino Library for the GAMMA function to adjust brightness of LED's etc.


## Description

This library is provides a gamma lookup class. It is typical used to 
change the intensity / brightness of a LED to match the human eye.
When a LED is given 50% PWM it looks far brighter for the eye.

This lib provides a balance between an expensive math function and the speed
of a lookup table. 

The accuracy of the library depends on the size of the internal array.
A size of 256 is the reference. Smaller arrays use interpolation and 
these interpolated values are good (< 1%) down to internal array size 16. 
The size can be as small as 2 which is pretty inaccurate.
In this latter case the curve is approximated by only two linear interpolations.
In short, choose the size that fits your application.

The library has a **setGamma(float gamma)** function that allows an application 
to change the gamma value runtime. 
This allows adjustments that are not possible with a fixed table.

The class provides **dump()** to create a table e.g. to place in PROGMEM. 
Since 0.2.2 the library also has **dumpArray()** to generate a C-style array.

Note: tested on UNO and ESP32 only.


## Interface

### Core functions

- **GAMMA(uint16_t size = 32)** constructor, gets the size of the internal
array as parameter. The default for size = 32 as this is a good balance between performance
and size of the internal array. 
The size parameter must be in {2, 4, 8, 16, 32, 64, 128, 256 }.
- **~GAMMA()** destructor.
- **bool begin()** The internal array is allocated and initialized with a gamma == 2.8.
This is an often used value to adjust light to human eye responses.
Note that **begin()** must be called before any other function.
Returns false if allocation fails.
- **void setGamma(float gamma)** calculates and fills the array with new values.
This can be done runtime so runtime adjustment of gamma mapping is possible.  
This calculation are relative expensive and takes quite some time (depending on size).
If the array already is calculated for gamma, the calculation will be skipped.
The parameter **gamma** must be > 0. The value 1 gives an 1:1 mapping.
Returns false if gamma <= 0 or if no table is allocated.
- **float getGamma()** returns the set gamma value.
- **uint8_t operator \[uint8_t index\]** allows the GAMMA object to be accessed as an array.
like ```x = G[40];``` Makes it easy to switch with a real array. 
The value returned is in the range 0 .. 255, so the user may need to scale it e.g. to 0.0 - 1.0.
Note: if index > size or internal table not allocated the function returns 0. 
As this is a legitimate value the user should take care.


### Development functions

- **uint16_t size()** returns size of the internal array.
This is always a power of 2.
- **uint16_t distinct()** returns the number of distinct values in the table. 
Especially with larger internal tables there will be duplicate numbers in the table.
- **bool dump(Stream \*str = &Serial)** dumps the internal table to a stream, default Serial. 
Useful to create an array in RAM, PROGMEM, EEPROM, in a file or wherever.
Returns false if no table is allocated.
- **void dumpArray(Stream \*str = &Serial)** dumps the internal table to a stream, default Serial, as a C-style array. See example.
Returns false if no table is allocated.


## Operation

See example.


## Future ideas

- improve documentation
- test other platforms
- look for optimizations
  - getter \[\]
  - setGamma -> pow() is expensive
- uint16 version?
  - GAMMA16, GAMMA32, 
  - GAMMA_RGB ?
- 



CHANGELOG.md
# Change Log Gamma

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).


## [0.4.0] - 2023-01-26
- optimize **setGamma()** for ESP32
- optimize **get[]** (minor)
- update GitHub actions
- update license 2023
- update readme.md

----

## [0.3.1] - 2022-11-08
- add changelog.md
- add rp2040 to build-CI
- update readme.md

## [0.3.0] - 2022-07-26
- change return type begin() + setGamma()
- add test gamma <=0 in setGamma()
- add _table == NULL tests
- fixed type of index in [] operator.
- adjust rounding in setGamma() to minimize errors.
- update build-CI

----

## [0.2.2   2022-07-25
- split in .h and .cpp
- add Stream parameter to dump()
- add dumpArray(Stream)
- fix distinct()

## [0.2.1   2021-12-18
- update library.json, license,
- add constants, minor edits.

## [0.2.0   2021-11-02
- update build-CI, badges
- add begin() - fixes ESP32 crash.

----

## [0.1.1] - 2020-12-24
- Arduino-CI + unit test

## [0.1.0] - 2020-08-08
- initial release



test/unit_test_001.cpp
//
//    FILE: unit_test_001.cpp
//  AUTHOR: Rob Tillaart
//    DATE: 2020-12-24
// PURPOSE: unit tests for the GAMMA library
//          https://github.com/RobTillaart/GAMMA
//          https://github.com/Arduino-CI/arduino_ci/blob/master/REFERENCE.md
//

// supported assertions
// https://github.com/Arduino-CI/arduino_ci/blob/master/cpp/unittest/Assertion.h#L33-L42
// ----------------------------
// assertEqual(expected, actual)
// assertNotEqual(expected, actual)
// assertLess(expected, actual)
// assertMore(expected, actual)
// assertLessOrEqual(expected, actual)
// assertMoreOrEqual(expected, actual)
// assertTrue(actual)
// assertFalse(actual)
// assertNull(actual)
// assertNotNull(actual)

#include <ArduinoUnitTests.h>


#include "Arduino.h"
#include "gamma.h"



unittest_setup()
{
  fprintf(stderr, "GAMMA_LIB_VERSION: %s\n", (char *) GAMMA_LIB_VERSION);
}

unittest_teardown()
{
}


unittest(test_constants)
{
  assertEqual( 32, GAMMA_DEFAULT_SIZE);
  assertEqual(256, GAMMA_MAX_SIZE);
}


unittest(test_constructor)
{
  GAMMA gt0;  // uses default 32 size
  assertTrue(gt0.begin());
  assertEqual(33, gt0.size());
  assertEqualFloat(2.8, gt0.getGamma(), 0.0001);
  assertEqual(29,  gt0.distinct());

  GAMMA gt1(256);
  assertTrue(gt1.begin());
  assertEqual(257, gt1.size());
  assertEqualFloat(2.8, gt1.getGamma(), 0.0001);
  assertEqual(163, gt1.distinct());

  GAMMA gt2(128);
  assertTrue(gt2.begin());
  assertEqual(129, gt2.size());
  assertEqualFloat(2.8, gt2.getGamma(), 0.0001);
  assertEqual(97,  gt2.distinct());

  GAMMA gt3(64);
  assertTrue(gt3.begin());
  assertEqual(65, gt3.size());
  assertEqualFloat(2.8, gt3.getGamma(), 0.0001);
  assertEqual(54,  gt3.distinct());

  GAMMA gt4(32);    // default
  assertTrue(gt4.begin());
  assertEqual(33, gt4.size());
  assertEqualFloat(2.8, gt4.getGamma(), 0.0001);
  assertEqual(29,  gt4.distinct());

  GAMMA gt5(16);
  assertTrue(gt5.begin());
  assertEqual(17, gt5.size());
  assertEqualFloat(2.8, gt5.getGamma(), 0.0001);
  assertEqual(16, gt5.distinct());

  GAMMA gt6(8);
  assertTrue(gt6.begin());
  assertEqual(9, gt6.size());
  assertEqualFloat(2.8, gt6.getGamma(), 0.0001);
  assertEqual(9, gt6.distinct());
}


unittest(test_get_set)
{
  GAMMA gt;  // uses default 32 size

  assertTrue(gt.begin());
  for (int i = 1; i < 20; i++)
  {
     assertTrue(gt.setGamma(i * 0.1));
     assertEqualFloat(i * 0.1, gt.getGamma(), 0.001);
  }
}


unittest(test_gamma_fail)
{
  GAMMA gt;  // uses default 32 size

  //  do not call begin() to force allocate error
  assertFalse(gt.setGamma(3.14);
  assertEqualFloat(1.0, gt.getGamma(), 0.001);
  
  assertEqualFloat(0.0, gt[63], 0.0001);
  assertFalse(gt.dump());
  assertFalse(gt.dumpArray());

  //  allocate
  assertTrue(gt.begin());
  //  still fails.
  assertFalse(gt.setGamma(-2.0));
  //  this works now
  assertTrue(gt.setGamma(3.14));
  assertEqualFloat(3.14, gt.getGamma(), 0.001);
}


unittest_main()

// --------


gamma.h
#pragma once
//
//    FILE: gamma.h
//  AUTHOR: Rob Tillaart
// VERSION: 0.3.0
//    DATE: 2020-08-08
// PURPOSE: Arduino Library to efficiently hold a gamma lookup table


#include "Arduino.h"

#define GAMMA_LIB_VERSION         (F("0.3.0"))

#define GAMMA_DEFAULT_SIZE        32
#define GAMMA_MAX_SIZE            256


class GAMMA
{

public:

  GAMMA(uint16_t size = GAMMA_DEFAULT_SIZE);
  ~GAMMA();

  //  allocates memory
  //  sets default gamma = 2.8
  //  Returns false if allocation fails
  bool begin();

  //  CORE
  //  Returns false if gamma <= 0
  bool  setGamma(float gamma);
  float getGamma();
  //  access values with index operator
  //  index = 0 .. size
  uint8_t operator[] (uint8_t index);

  //  META INFO
  uint16_t size();
  uint16_t distinct();

  //  DEBUG
  bool dump(Stream *str = &Serial);
  bool dumpArray(Stream *str = &Serial);


private:
  uint8_t   _shift    = 0;
  uint8_t   _mask     = 0;
  uint16_t  _size     = 0;
  uint8_t   _interval = 0;
  float     _gamma    = 0;
  uint8_t * _table    = NULL;

  float     fastPow(float a, float b);
};


// -- END OF FILE --



gamma.cpp
//
//    FILE: gamma.cpp
//  AUTHOR: Rob Tillaart
// VERSION: 0.3.0
//    DATE: 2020-08-08
// PURPOSE: Arduino Library to efficiently hold a gamma lookup table

//  0.1.0   2020-08-08  initial release
//  0.1.1   2020-12-24  Arduino-CI + unit test
//
//  0.2.0   2021-11-02  update build-CI, badges
//                      add begin() - fixes ESP32 crash.
//  0.2.1   2021-12-18  update library.json, license, 
//                      add constants, minor edits.
//  0.2.2   2022-07-25  split in .h and .cpp
//                      add Stream parameter to dump()
//                      add dumpArray(Stream)
//                      fix distinct()
//
//  0.3.0   2022-07-26  change return type begin + setGAMMA
//                      add test gamma <=0 in setGamma
//                      add _table == NULL tests
//                      fixed type of index in [] operator.
//                      adjust rounding in setGamma to minimize errors.
//                      update build-CI


#include "gamma.h"


GAMMA::GAMMA(uint16_t size)
{
  _shift = 7;
  // force power of 2; get shift & mask right
  for (uint16_t s = 2; s <= GAMMA_MAX_SIZE; s <<= 1)
  {
    if (size <= s)
    {
      _size = s;
      break;
    }
    _shift--;
  }
  _mask = (1 << _shift) - 1;
  _interval = GAMMA_MAX_SIZE / _size;
};


GAMMA::~GAMMA()
{
  if (_table) free(_table);
};


bool GAMMA::begin()
{
  if (_table == NULL)
  {
    _table = (uint8_t *)malloc(_size + 1);
  }
  if (_table == NULL) return false;
  setGamma(2.8);
  return true;
};


bool GAMMA::setGamma(float gamma)
{
  if (_table == NULL) return false;
  if (gamma <= 0) return false;
  if (_gamma != gamma)
  {
    yield();  // try to keep ESP happy
    _gamma = gamma;
    //  marginally faster
    // uint16_t iv = _interval;
    // _table[0] = 0;
    // for (uint16_t i = 1; i < _size; i++)
    // {
      // float x = log(i * iv) + log(1.0 / 255);
      // _table[i] = exp(x * _gamma) * 255 + 0.5;
    // }
    //  REFERENCE
    //  rounding factor 0.444 optimized with error example.
    for (uint16_t i = 1; i < _size; i++)
    {
      _table[i] = pow(i * _interval * (1.0/ 255.0), _gamma) * 255 + 0.444;
    }
    _table[0] = 0;
    _table[_size] = 255;  // anchor for interpolation.
  }
  return true;
};


float GAMMA::getGamma()
{
  return _gamma;
};


uint8_t GAMMA::operator[] (uint8_t index)
{
  //  0.3.0 _table test slows performance ~0.4 us.
  if (_table == NULL) return 0;
  if (_interval == 1) return _table[index];
  // else interpolate
  uint8_t  i = index >> _shift;
  uint8_t  m = index & _mask;
  // exact element shortcut
  if ( m == 0 ) return _table[i];
  //  interpolation
  //  delta must be uint16_t to prevent overflow. (small tables)
  //        delta * m can be > 8 bit.
  uint16_t delta = _table[i+1] - _table[i];
  delta = (delta * m + _interval/2) >> _shift;  //  == /_interval;
  return _table[i] + delta;
};


uint16_t GAMMA::size()
{
  return _size + 1;
};


uint16_t GAMMA::distinct()
{
  uint8_t  last  = _table[0];
  uint16_t count = 1;
  for (uint16_t i = 1; i <= _size; i++)
  {
    if (_table[i] == last) continue;
    last = _table[i];
    count++;
  }
  return count;
};


bool GAMMA::dump(Stream *str)
{
  if (_table == NULL) return false;
  for (uint16_t i = 0; i <= _size; i++)
  {
    str->println(_table[i]);
  }
  return true;
};


bool GAMMA::dumpArray(Stream *str)
{
  if (_table == NULL) return false;
  str->println();
  str->print("uint8_t gamma[");
  str->print(_size + 1);
  str->print("] = {");
  
  for (uint16_t i = 0; i <= _size; i++)
  {
    if (i % 8 == 0) str->print("\n  ");
    str->print(_table[i]);
    if (i < _size) str->print(", ");
  }
  str->print("\n  };\n\n");
  return true;
};


//  performance investigation
//  https://stackoverflow.com/questions/43429238/using-boost-cpp-int-for-functions-like-pow-and-rand
inline float GAMMA::fastPow(float a, float b)
{
  //  reference
  return pow(a, b);
}


// -- END OF FILE --

